client/json-rpc/src/async_client/client.rs

```diff
@@ -248,85 +348,175 @@ impl<R: RetryStrategy> Client<R> {
+    /// let currencies: Response<jsonrpc::CurrencyInfo> = res
+    ///     .remove(0)
+    ///     .try_into()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// For get_account and get_account_transaction request, the above convert will return
+    /// result not found error instead of None.
+    ///
+    /// This function calls to batch_send_with_retry, when you batch submit transactions,
+    /// to avoid re-submit errors, it is better to use batch_send_without_retry and handle errors
+    /// by yourself.
+    pub async fn batch_send(&self, requests: Vec<Request>) -> Result<Vec<JsonRpcResponse>, Error> {
+        self.batch_send_with_retry(&requests).await
+    }
+
+    pub async fn batch_send_with_retry(
         &self,
-        method: &str,
-        params: serde_json::Value,
-    ) -> Result<JsonRpcResponse, Error> {
+        requests: &[Request],
+    ) -> Result<Vec<JsonRpcResponse>, Error> {
         let mut retries: u32 = 0;
         loop {
-            let ret = self.send_without_retry(method, &params).await;
+            let ret = self.batch_send_without_retry(requests).await;
             match ret {
-                Ok(r) => {
-                    return Ok(r);
-                }
-                Err(err) => {
-                    if !self.retry.is_retriable(&err) {
-                        return Err(err);
-                    }
-                    if retries < self.retry.max_retries(&err) {
-                        match retries.checked_add(1) {
-                            Some(i) if i < self.retry.max_retries(&err) => {
-                                retries = i;
-                            }
-                            _ => return Err(err),
-                        };
-                        tokio::time::delay_for(self.retry.delay(&err, retries)).await;
-                        continue;
-                    }
-                    return Err(err);
-                }
+                Ok(r) => return Ok(r),
+                Err(err) => retries = self.handle_retry_error(retries, err).await?,
             }
         }
     }
 
-    pub async fn send_without_retry(
+    pub async fn batch_send_without_retry(
         &self,
-        method: &str,
-        params: &serde_json::Value,
-    ) -> Result<JsonRpcResponse, Error> {
+        requests: &[Request],
+    ) -> Result<Vec<JsonRpcResponse>, Error> {
+        let json_requests: Vec<serde_json::Value> = requests
+            .iter()
+            .enumerate()
+            .map(|(i, r)| r.to_json(i))
+            .collect();
+
+        let rpc_resps: Vec<JsonRpcResponse> = self.send_json_request(json![json_requests]).await?;
+        if rpc_resps.len() != requests.len() {
+            return Err(Error::unexpected_invalid_batch_response(rpc_resps));
+        }
+
+        let mut result: Vec<JsonRpcResponse> = rpc_resps.clone();
+        for resp in rpc_resps {
+            let id = self.validate(&resp, 0, requests.len() - 1)?;
+            result[id] = resp;
```

Comment:
this won't catch issues where two response have same id . I still think the logic would be easier if you were to to use an hashmap then fill result by looking up id as key

```
resp_maps = HashMap::new() 
for resp in. rpc_resps {
 id = self.validate(0, requests.len() -1)?
 if resp_maps.contains(id) {
    // error
 } 
}
for i in 0..requests.len()-1 {
   result.append(resp_maps.get(i)?)
}
```